pkg/kubelet/pod_workers.go

```diff
@@ -599,9 +602,18 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 	}
 
 	// start the pod worker goroutine if it doesn't exist
-	var podUpdates chan podWork
-	var exists bool
-	if podUpdates, exists = p.podUpdates[uid]; !exists {
+	podUpdates, ok := p.podUpdates[uid]
+	if !ok {
+		if kubetypes.IsStaticPod(pod) {
+			fullname := kubecontainer.GetPodFullName(pod)
+			if _, exist := p.managedStaticPodFullnames[fullname]; exist {
+				klog.InfoS("Pod update is reserved, wait for outdated mirror pod to be terminated", "pod", klog.KObj(pod), "podUID", pod.UID)
+				p.workQueue.Enqueue(uid, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
```

Comment:
I think I'd prefer to handle this by having the worker complete the previous work and restart the container in the new state inside the pod worker loop.   But the real problem is that the kubelet has two completely different systems for tracking uniqueness - apiservers use UID (which technically could be duplicated), and static pod use name and namespace AND UID - and I'm not even sure this protects against that.

The pod worker needs to keep track of all in flight admitted pods.  I'm at least at first glance uncomfortable with the pod worker dropping a known pod.  Instead, it might be better to have the pod worker handle both cases of:

1. pod changes name/namespace but not UID
2. pod changes UID but not name/namespace

as a "pending pod" (static pods probably need priority) within the worker, and have the worker decide which of them starts after a running pod completes.  No matter what the pod worker has to keep track of previously running pods anyway.

So roughly (there are some problems here that need more thought):

1. on UpdatePod identify whether there are name/namespace or uid conflicts on "add/update" with already running pods
2. if there is and the pod is already terminating, queue that pod for when the worker completes
3. if there is and the pod is not already terminating AND the incoming pod is a static pod and the existing one is not, we probably need to create a synthetic UID for that net new pod and start it

We can't rely on the kubelet config loop ensuring a unique UID, because neither API server nor static pod can guarantee that, and on a restart we may still have running pods that have to be terminated before the new version can start.

Reply:
The next concern I have in this is also that we're enqueuing to the work queue from this method.  In practice I'd prefer that this method only start workers, and if we need to wake another queued worker we do it from the state change methods inside the worker.  

Outside of that, I think we're trying to queue static pods when a) another static pod has the same full name but a different uid and b) when another pod has the same UID and is not a static pod.  If it has a different UID, we need to start a worker for it, but ensure that it can't start until the previous is cleared.

Reply:
one more note: we're trying to preserve the property that all valid accepted static pods start - so we can't block a static pod from starting if a non-static-pod has the same name, but we can make it so that with two static pods with the same uid correctly block each other, and that a static pod that reuses the same uid as a real pod waits for the other pod to terminate (kubelet config guarantees only one pod with a given uid is accepted at any time, and each kubelet config source guarantees that no pod within the same source has the same name at any time).

Part of why I'm reviewing this more thoroughly is that in order to pod admission to properly reflect pods that are terminating after force deletion (#103979) we will need to keep track of all current and admitted pods in the pod worker, which means we're about to add even more tracking in the pod worker.

Reply:
https://github.com/kubernetes/kubernetes/issues/104648 looks related

the calculated uid includes the source filename, so two static pods from different files would have different uids

Reply:
@smarterclayton 
Do we have to consider two different pods with the same uid?

I think
- Pod uids are unique inside of each source(apiserver, file, and HTTP).
- static pods and non-static pods have different uid formats(non-static pod's uid is UUID format...?)

Let me know if I misunderstand what you said.

Reply:
> as a "pending pod" (static pods probably need priority) within the worker, and have the worker decide which of them starts after a running pod completes.

Do we have to keep track of all pending static pods? 
I am planning to reserve one most recent static pod for each pod worker.